std/selector/slice.go

```diff
@@ -0,0 +1,87 @@
+package selector
+
+import (
+	"github.com/consensys/gnark/constraint/solver"
+	"github.com/consensys/gnark/frontend"
+	"math/big"
+)
+
+func init() {
+	// register hints
+	solver.RegisterHint(StepOutput)
+}
+
+// Partition selects left or right side of the input array, with respect to the pivotPosition.
+// More precisely when rightSide is false, for each i we have:
+//
+//	if i < pivotPosition
+//	    out[i] = input[i]
+//	else
+//	    out[i] = 0
+//
+// and when rightSide is true, for each i we have:
+//
+//	if i >= pivotPosition
+//	    out[i] = input[i]
+//	else
+//	    out[i] = 0
+//
+// We must have pivotPosition >= 1 and pivotPosition <= len(input)-1, otherwise a proof cannot be generated.
+func Partition(api frontend.API, pivotPosition frontend.Variable, rightSide bool,
```

Comment:
Looks good.

What do you think about having additionally `Slice` method a la
```go
func Slice(api frontend.API, start, end frontend.Variable, input []frontend.Variable) ([]frontend.Variable) {
    out := Partition(api, start, true, input)
    out = Partition(api, end, false, out)
    return out
}
```

Imo it is more natural to the way people use slices in Go as `out := in[start:end]`.

Additionally, maybe we can even have modified methods `PartitionCut` and `SliceCut` with additional argument `outPutlen` which returns an array of length `outPutlen` (int type), omitting the zeros. So for example:
```go
in := []frontend.Variable{1,2,3,4,5}
out := SliceCut(api, 1, 3, in, 2) // == []frontend.Variable{2,3}
out = SliceCut(api, 1, 3, in, 3) // == []frontend.Variable{2,3,0} as outputLen is larger than end-start
out = SliceCut(api, 1, 3, in, 1) // == []frontend.Variable{2}, result incomplete
out = SliceCut(api, 1, 3, in, 6) // panic as outLen > len(in)
```

Reply:
But I guess if there is a `Slice` function then we can do it slightly more efficiently when we do it in a single pass instead of two in my proposal.

Reply:
Actually, I think we should implement a `Slice` method. The thing is, after a while I realized that cascading is not a good idea and working with `mask` and using addition and subtraction can be more efficient. That's why I made `StepMask` public (exported). For example:
```
mask1         = [0, 0, 0, 1, 1, 1, 1, 1]
mask2         = [0, 0, 0, 0, 0, 0, 1, 1]
mask1 - mask2 = [0, 0, 0, 1, 1, 1, 0, 0]
```
So we will avoid an array multiplication.

I think implementing `SliceCut` won't be so easy, since `end` and `start` are wires. We would need to use `Mux` to *shift* every element.

Reply:
It's tricky, I'll try to explain in the next comment.

Reply:
I agree. It seems tricky. Lets omit `..Cut` versions.